LSLib/Granny/ClothUtils.cs

```diff
@@ -0,0 +1,384 @@
+        {
+            var clothVertices = new HashSet<ClothVertex>(mesh.Vertices.Where(v => v.Weight > 0));
+
+            AddOverlappingVertices(clothVertices);
+            AddNeighboringVertices(clothVertices);
+            AddOverlappingVertices(clothVertices);
+            AddNeighboringVertices(clothVertices);
+            AddOverlappingVertices(clothVertices);
+
+            return clothVertices;
+        }
+
+        private static void AddOverlappingVertices(HashSet<ClothVertex> clothVertices)
+        {
+            foreach (var vertex in clothVertices.SelectMany(v => v.Overlapping).ToList())
+            {
+                clothVertices.Add(vertex);
+            }
+        }
+
+        private static void AddNeighboringVertices(HashSet<ClothVertex> clothVertices)
+        {
+            foreach (var vertex in clothVertices.SelectMany(v => v.Neighbors).ToList())
+            {
+                clothVertices.Add(vertex);
+            }
+        }
+
+        private static unsafe short GetShort(byte[] array, int offset)
+        {
+            fixed (byte* ptr = &array[offset])
+            {
+                return *(short*)ptr;
+            }
+        }
+
+        private static unsafe void SetBytes(short value, byte[] array, int offset)
+        {
+            fixed (byte* ptr = &array[offset])
+            {
+                *(short*)ptr = value;
+            }
+        }
+
+        private record ClothMesh
+        {
+            private ClothMesh(IReadOnlyList<ClothVertex> vertices, IReadOnlyList<int> indices)
+            {
+                Vertices = vertices;
+                Indices = indices;
+            }
+
+            internal IReadOnlyList<ClothVertex> Vertices { get; }
+
+            internal IReadOnlyList<int> Indices { get; }
+
+            internal static ClothMesh Build(Mesh mesh)
+            {
+                var vertices = new ClothVertex[mesh.PrimaryVertexData.Vertices.Count];
+
+                for (int i = 0; i < vertices.Length; i++)
+                {
+                    vertices[i] = new ClothVertex(i, mesh.PrimaryVertexData.Vertices[i]);
+                }
+
+                var kdTree = BuildKdTree(vertices);
+
+                foreach (var vertex in vertices)
+                {
+                    // searchRadius is squared because our tree uses the square distance as its metric
+                    var overlapping = kdTree.RadialSearch(vertex.PositionAsArray, OverlappingVertexSearchRadius * OverlappingVertexSearchRadius).Select(v => v.Item2).OrderBy(v => v.Index).ToList();
+
+                    // nothing to do if there are no overlapping vertices (we just found the vertex itself)
+                    if (overlapping.Count == 1)
+                    {
+                        continue;
+                    }
+
+                    vertex.Weight = overlapping[0].Weight;
+                    
+                    foreach (var other in overlapping)
```

Comment:
may be simpler as `vertex.Overlapping.AddRange(overlapping.Where(v => v != other))`;